core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -314,6 +288,21 @@ private static long optionalSum(long left, long right) {
         return scoreDocs;
     }
 
+    static <T extends TopDocs> void fillTopDocs(T[] shardTopDocs,
+                                                        List<? extends AtomicArray.Entry<? extends QuerySearchResultProvider>> results,
+                                                        Supplier<T> emptySupplier) {
+        if (results.size() != shardTopDocs.length) {
+            // TopDocs#merge can't deal with null shard TopDocs
+            final T empty = emptySupplier.get();
```

Comment:
the supplier seems to be only useful to save one object allocation, maybe getting rid of it would make things simpler?